src/openvic-simulation/economy/Producer.hpp

```diff
@@ -0,0 +1,86 @@
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+#include <map>
+
+#include "openvic-simulation/economy/Good.hpp"
+#include "openvic-simulation/economy/ProductionType.hpp"
+#include "openvic-simulation/types/fixed_point/FixedPoint.hpp"
+#include "openvic-simulation/utility/Getters.hpp"
+
+using money_t = OpenVic::fixed_point_t;
+using good_quantity_t = OpenVic::fixed_point_t;
+
+namespace OpenVic {
+	class Producer {
+	public:
+		ProductionType PROPERTY(production_type)
+	};
+
+	class Manufacturer : Producer {
+	public:
+		Good::good_map_t PROPERTY(stockpile);
+	};
+
+	class ArtisanalProducer final : Manufacturer {
+	public:
+		good_quantity_t PROPERTY(current_production);
+		Good::good_map_t PROPERTY(current_needs);
+	};
+
+	class NonArtisanalProducer : Producer {
+	private:
+		Pop::pop_size_t get_max_workforce_size() {
+			return static_cast<Pop::pop_size_t>(
+				static_cast<fixed_point_t>(production_type.get_base_workforce_size()) * size_multiplier
+			);
+		}
+
+	public:
+		money_t PROPERTY(revenue_yesterday);
+		good_quantity_t PROPERTY(output_quantity_yesterday);
+		good_quantity_t PROPERTY(unsold_quantity_yesterday);
+		fixed_point_t PROPERTY(size_multiplier);
+		std::map<Pop const*, Pop::pop_size_t> PROPERTY(employees)
+	};
+
+	class FactoryProducer final : NonArtisanalProducer, Manufacturer {
+	private:
+		int PROPERTY(profit_history_current);
+		money_t[7] PROPERTY(daily_profit_history);
+
+		fixed_point_t get_max_budget() {
+			return defines.economy.MAX_FACTORY_MONEY_SAVE * size_multiplier;
+		}
+
+	public:
+		money_t PROPERTY(budget);
+		money_t PROPERTY(balance_yesterday);
+		money_t PROPERTY(received_investments_yesterday);
+		money_t PROPERTY(market_spendings_yesterday);
+		money_t PROPERTY(paychecks_yesterday);
+		uint32_t PROPERTY(unprofitable_days);
+		uint32_t PROPERTY(injected_days);
+		uint32_t PROPERTY(days_without_input);
+		std::byte PROPERTY_RW_ACCESS(hiring_priority, public);
+
+		fixed_point_t get_profitability_yesterday() {
+			return daily_profit_history[profit_history_current];
+		}
+
+		fixed_point_t get_average_profitability_last_seven_days() {
+			money_t sum = 0;
+
+			for (int i = 0; i <= profit_history_current; i++) {
+				sum += daily_profit_history[i];
+			}
+
+			return sum / (1 + profit_history_current);
+		}
+	};
+
+	class ResourceGatheringOpertion final : NonArtisanalProducer {
+		ResourceGatheringOpertion(const ProductionType production_type, const fixed_point_t size_multiplier);
```

Comment:
As mention in `Producer`, `production_type` should be a `ProductionType const*`. Also function arguments never need to be `const` (pointers and references can need `const` before the `*`/`&`, but that's different from the variable itself being const). In the header definition `const` arguments are extra unnecessary, it doesn't actually have any effect, the implementation's arguments don't have to match the `const`-ness of the header definition's arguments.